Update ophyd_tools.py

awalter-bnl opened the pull request:
This PR is to allow for the inclusion of 'positions' that return a string instead of a float. It gets around the fact that it was hard-coded to apply 'round' to the value prevented this.

awalter-bnl committed:
Update ophyd_tools.py

tacaswell commented on pyOlog/ophyd_tools.py:
```diff
@@ -445,11 +445,14 @@ def _print_pos(positioners, file=sys.stdout):
+                    prec = p.precision
+                except (AttributeError, DisconnectedError):
+                    prec = FMT_PREC
+                value = np.round(v, decimals=prec)
```
This might fail on some weird edge cases due to duck-typing.

I suggest

```python
try:
    value = np.round(v, decimals=prec)
except TypeError:
    value = v
```

awalter-bnl replied:
Thanks @tacaswell , I was not aware of the weird edge cases. I will make the change you suggest.

awalter-bnl committed:
Change from using 'if isinstance(v,flost)......' to using 'try...except TypeError: .......' as sugested by Tom.

tacaswell commented on pyOlog/ophyd_tools.py:
```diff
@@ -446,10 +446,12 @@ def _print_pos(positioners, file=sys.stdout):
-            except (AttributeError, DisconnectedError):
-                prec = FMT_PREC
-            value = np.round(v, decimals=prec)
+                value = np.round(v, decimals=prec)
```
This is going to raise an NameError.  We need two try/excepts here, one to cover (maybe) getting the precision from the PV and one to cover (maybe) rounding it.

awalter-bnl replied:
I am not sure I follow your comment here, How can we seperate out the 'tries' as they are both errors from the same process (i.e. trying value = np.round(v, decimals=prec) ). 

One case is a type error (i.e. v is a string) and the second is an attribute error/ disconnected error where prec is not defined.   How do you suggest we split this into 2 tries?

tacaswell replied:
```python
try:
    perc = p.precision
except (AttributeError, DisconnectedError):
    perc = FMT_PERC

try:
    value = np.round(v, decimals=prec)
except TypeError:
    value = v
```

awalter-bnl replied:
OK, I am not seeing how this will result in a different outcome. but I am sure it is my own ignorance !-) I will play around myself and see why it won't work, in the meantime I will make this change.

mrakitin commented on pyOlog/ophyd_tools.py:
```diff
@@ -446,10 +446,12 @@ def _print_pos(positioners, file=sys.stdout):
+                value = np.round(v, decimals=prec)
+            except TypeError:
+                value = v
+            except (AttributeError,DisconnectedError):
```
Do we care about PEP8 in this project? I mean missed blank space.

awalter-bnl replied:
I will fix this, just waiting on some feedback from Tom on his comment

awalter-bnl committed:
Update Ophyd Tools, log_pos

splitting an exisitng try...... except....except into try........except, try..........except

awalter-bnl committed:
Update ophyd_tools.py

mrakitin commented on pyOlog/ophyd_tools.py:
```diff
@@ -196,10 +196,15 @@ def log_pos(positioners=None, extra_msg=None):
+       try:
+           prec = p.precision
+       except (AttributeError, DisconnectedError):
+           prec = FMT_PERC
```
I think it's a typo here, should match the one you use later (`FMT_PREC`).

awalter-bnl replied:
FIxed this, sorry

awalter-bnl committed:
Update ophyd_tools.py

tacaswell commented on pyOlog/ophyd_tools.py:
```diff
@@ -196,10 +196,15 @@ def log_pos(positioners=None, extra_msg=None):
-            pdict['values'][p.name] = p.position
-        except DisconnectedError:
-            pdict['values'][p.name] = DISCONNECTED
+       try:
```
I don't think this block should have been modified.

awalter-bnl replied:
Wow, sorry I put the new code in the wrong place.  I am not sure how I did that. I will resolve it now

awalter-bnl committed:
Update ophyd_tools.py

awalter-bnl committed:
Update ophyd_tools.py

awalter-bnl committed:
Update ophyd_tools.py
